File: examples/access/tessellation/tile_properties.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Access tile metadata properties after tessellation.
"""

import numpy as np
import brahe as bh

bh.initialize_eop()

# ISS TLE
line1 = "1 25544U 98067A   08264.51782528 -.00002182  00000-0 -11606-4 0  2927"
line2 = "2 25544  51.6416 247.4627 0006703 130.5360 325.0288 15.72125391563537"

# Create propagator and tessellator
prop = bh.SGPPropagator.from_tle(line1, line2, step_size=60.0)
config = bh.OrbitGeometryTessellatorConfig(
    image_width=5000,
    image_length=5000,
    asc_dsc=bh.AscDsc.ASCENDING,
)
tess = bh.OrbitGeometryTessellator(prop, prop.epoch, config, spacecraft_id="ISS")

# Tessellate a point and inspect properties
point = bh.PointLocation(10.0, 30.0, 0.0)
tiles = tess.tessellate_point(point)
tile = tiles[0]
props = tile.properties

# Along-track direction (unit vector in ECEF)
direction = np.array(props["tile_direction"])
print(f"tile_direction: [{direction[0]:.4f}, {direction[1]:.4f}, {direction[2]:.4f}]")
print(f"  magnitude: {np.linalg.norm(direction):.6f}")

# Tile dimensions
print(f"tile_width: {props['tile_width']:.0f} m")
print(f"tile_length: {props['tile_length']:.0f} m")
print(f"tile_area: {props['tile_area']:.0f} m^2")

# Group and spacecraft identifiers
print(f"tile_group_id: {props['tile_group_id'][:8]}...")
print(f"spacecraft_ids: {props['spacecraft_ids']}")

File: examples/access/tessellation/tile_properties.rs
//! Access tile metadata properties after tessellation.
//! Demonstrates reading direction, dimensions, and identifiers from tiles.

#[allow(unused_imports)]
use brahe as bh;
use bh::access::constraints::AscDsc;
use bh::access::location::{AccessibleLocation, PointLocation};
use bh::access::tessellation::{OrbitGeometryTessellator, OrbitGeometryTessellatorConfig, Tessellator};
use bh::propagators::SGPPropagator;

fn main() {
    bh::initialize_eop().unwrap();

    // ISS TLE
    let line1 = "1 25544U 98067A   08264.51782528 -.00002182  00000-0 -11606-4 0  2927";
    let line2 = "2 25544  51.6416 247.4627 0006703 130.5360 325.0288 15.72125391563537";

    // Create propagator and tessellator
    let prop = SGPPropagator::from_tle(line1, line2, 60.0).unwrap();
    let epoch = prop.epoch;
    let config = OrbitGeometryTessellatorConfig::default()
        .with_asc_dsc(AscDsc::Ascending);
    let tess = OrbitGeometryTessellator::new(
        Box::new(prop), epoch, config, Some("ISS".to_string()),
    );

    // Tessellate a point and inspect properties
    let point = PointLocation::new(10.0, 30.0, 0.0);
    let tiles = tess.tessellate(&point).unwrap();
    let props = tiles[0].properties();

    // Along-track direction (unit vector in ECEF)
    let dir = props["tile_direction"].as_array().unwrap();
    let dx = dir[0].as_f64().unwrap();
    let dy = dir[1].as_f64().unwrap();
    let dz = dir[2].as_f64().unwrap();
    let mag = (dx * dx + dy * dy + dz * dz).sqrt();
    println!("tile_direction: [{:.4}, {:.4}, {:.4}]", dx, dy, dz);
    println!("  magnitude: {:.6}", mag);

    // Tile dimensions
    println!("tile_width: {:.0} m", props["tile_width"].as_f64().unwrap());
    println!("tile_length: {:.0} m", props["tile_length"].as_f64().unwrap());
    println!("tile_area: {:.0} m^2", props["tile_area"].as_f64().unwrap());

    // Group and spacecraft identifiers
    let group_id = props["tile_group_id"].as_str().unwrap();
    println!("tile_group_id: {}...", &group_id[..8]);
    println!("spacecraft_ids: {}", props["spacecraft_ids"]);
}

